### PR TITLE
Replace pyutilib.misc.import_file

### PR DESCRIPTION
**Pull request by mrmundt**
## Summary/Motivation:
A Pyomo 6.0.0 initiative is to remove the PyUtilib package. This replaces `pyutilib.misc.import_file` with a simpler implementation (using new Python 3 libraries) in `pyomo.common.fileutils`.

## Changes proposed in this PR:
- Remove `pyutilib.misc.import_file`
- Make new implementation in `pyomo.common.fileutils`

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/pysp/tests/convert/test_smps_embedded.py [hunk 1]
```diff
@@ -34,6 +33,7 @@
                                    StageCostAnnotation,
                                    VariableStageAnnotation)
 from pyomo.pysp.convert.smps import convert_embedded
+from pyomo.common.fileutils import import_file
```
Comment (jsiirola) [inline pyomo/pysp/tests/convert/test_smps_embedded.py:36]: Should we be making any changes in PySP?

File: pyomo/pysp/util/misc.py [hunk 2]
```diff
@@ -21,6 +21,7 @@
 import argparse
 
 from pyutilib.misc import import_file
+
```
Comment (jsiirola) [inline pyomo/pysp/util/misc.py:24]: Should we be making any changes in PySP?

File: pyomo/pysp/scenariotree/manager.py [hunk 3]
```diff
@@ -23,10 +23,10 @@
 from collections import (defaultdict,
                          namedtuple)
 
-import pyutilib.misc
 from pyutilib.pyro import (shutdown_pyro_components,
                            using_pyro4)
 from pyomo.common.dependencies import dill, dill_available
+from pyomo.common.fileutils import import_file
```
Comment (jsiirola) [inline pyomo/pysp/scenariotree/manager.py:29]: Should we be making any changes in PySP?

File: pyomo/pysp/phsolverserver.py [hunk 4]
```diff
@@ -969,8 +969,7 @@ def invoke_external_function(self,
         elif module_name in sys.modules:
             this_module = sys.modules[module_name]
         else:
-            this_module = import_file(module_name,
-                                      clear_cache=True)
+            this_module = import_file(module_name)
```
Comment (jsiirola) [inline pyomo/pysp/phsolverserver.py:972]: Should we be making any changes in PySP?

File: pyomo/pysp/phinit.py [hunk 5]
```diff
@@ -15,11 +15,11 @@
 import argparse
 
 from pyutilib.pyro import shutdown_pyro_components
-from pyutilib.misc import import_file
 
 from pyomo.common import pyomo_command
 from pyomo.common.dependencies import pympler_available, attempt_import
 from pyomo.common.plugin import ExtensionPoint, SingletonPlugin
+from pyomo.common.fileutils import import_file
```
Comment (jsiirola) [inline pyomo/pysp/phinit.py:22]: Should we be making any changes in PySP?

File: pyomo/contrib/pynumero/tests/test_cyipopt_examples.py [hunk 6]
```diff
@@ -36,13 +45,13 @@ def test_status_maps(self):
 
 class TestExamples(unittest.TestCase):
     def test_external_grey_box_react_example_maximize_cb_outputs(self):
-        ex = import_file(os.path.join(example_dir, 'external_grey_box', 'react-example', 'maximize_cb_outputs.py'))
+        ex = import_file(os.path.join(example_dir, 'external_grey_box', 'react_example', 'maximize_cb_outputs.py'))
```
Comment (jsiirola) [inline pyomo/contrib/pynumero/tests/test_cyipopt_examples.py:48]: Does `import_file()` not work if there are "`-`" in the filenames / paths?

**Review by jsiirola — COMMENTED** [review]
Generally, this looks good.  Two big questions, though:
- does import_file no longer work if there are dashes (or spaces) in the path?  That may be a deal-breaker.
- Should we be changing anything in PySP?

File: pyomo/contrib/pynumero/tests/test_cyipopt_examples.py (hunk 6 above)
Comment (mrmundt) [reply pyomo/contrib/pynumero/tests/test_cyipopt_examples.py:48]: The issue here wasn't `import_file` not working - it was a relative import at the top of the file that was causing issues. `import_file` doesn't care.

File: pyomo/common/tests/test_fileutils.py [hunk 7]
```diff
@@ -99,6 +100,19 @@ def test_this_file_dir(self):
         expected_path = os.path.join('pyomo','common','tests')
         self.assertTrue(_this_file_dir.endswith(expected_path))
 
+    def test_import_file(self):
+        import_ex = import_file(os.path.join(_this_file_dir, 'import_ex.py'))
```
Comment (jsiirola) [inline pyomo/common/tests/test_fileutils.py:104]: Out of curiosity, what happens if the file is `import ex.py` or `import-ex.py`?

Comment (mrmundt) [reply pyomo/common/tests/test_fileutils.py:104]: Tested it in ipython with a file `import-ex.py` in a folder `test-this`:
```
In [6]: test_dir = os.path.join(os.getcwd(), 'test-this/import-ex.py')

In [7]: os.path.isfile(test_dir)
Out[7]: True

In [8]: from pyomo.common.fileutils import import_file

In [9]: module = import_file(test_dir)

In [10]: module
Out[10]: <module 'import-ex' from '/Users/mmundt/Documents/devel-pyomo/mmundt-pyomo/test-this/import-ex.py'>

In [11]: 
```

Comment (mrmundt) [reply pyomo/common/tests/test_fileutils.py:104]: I also tried with the file `import ex.py` in `test-this`:
```
In [1]: from pyomo.common.fileutils import import_file

In [2]: import os

In [3]: test_dir = os.path.join(os.getcwd(), 'test-this/import ex.py')

In [4]: os.path.isfile(test_dir)
Out[4]: True

In [5]: module = import_file(test_dir)

In [6]: module
Out[6]: <module 'import ex' from '/Users/mmundt/Documents/devel-pyomo/mmundt-pyomo/test-this/import ex.py'>
```

Comment (mrmundt) [reply pyomo/common/tests/test_fileutils.py:104]: Do you want me to add these cases to the tests?

**Review by jsiirola — APPROVED** [review]

File: pyomo/scripting/util.py [hunk 8]
```diff
@@ -502,6 +505,8 @@ def apply_optimizer(data, instance=None):
     if len(data.options.solvers[0].suffixes) > 0:
         for suffix_name in data.options.solvers[0].suffixes:
             if suffix_name[0] in ['"',"'"]:
+                # !!THIS SEEMS LIKE A BUG!! - mrmundt #
+                # Suffix has not yet been defined.
                 suffix_name = suffix[1:-1]
```
Comment (blnicho) [inline pyomo/scripting/util.py:510]: I think `suffix` should be `suffix_name` here

File: pyomo/common/fileutils.py [hunk 9]
```diff
@@ -388,6 +390,34 @@ def find_executable(exename, cwd=True, include_PATH=True, pathlist=None):
...
+        raise Exception('File does not exist. Check path.')
+    try:
+        module_dir, module_file = os.path.split(path)
+        module_name, module_ext = os.path.splitext(module_file)
+        if clear_cache and module_name in sys.modules:
+            del sys.modules[module_name]
+        spec = importlib.util.spec_from_file_location(module_name, path)
+        module = spec.loader.load_module()
+    except:
+        raise
```
Comment (jsiirola) [inline pyomo/common/fileutils.py:417]: If we are going to leave this like this, then we should remove the try-except block entirely.  Personally, I think this is a good change, as suppressing the ImportError (or the SyntaxError) made debugging modules imported through import_file rather challenging.  However, we will need to go through all the places where import_file is called and (if necessary) replace the checks for `None` with handling of the raised exception.

**Review by jsiirola — CHANGES_REQUESTED** [review]
A few nit-picky questions around the `import_file` implementation, then I think this will be good to go.

File: pyomo/common/fileutils.py (hunk 9 above)
Comment (mrmundt) [reply pyomo/common/fileutils.py]: I did not find any checks for `None`. Other changes have been implemented.

**Review by jsiirola — APPROVED** [review]

**Review by blnicho — APPROVED** [review]

**Pull request merged**